### PR TITLE
Add LTF-inspired drift injection functions

### LINKED ISSUE #125: Synthetic noise injection for drift detector parameter selection
- Related to #123. 
- Realistic drift scenarios probably rely heavily on domain knowledge for the data.
- There's a lot of related work in robustness/sensitivity analysis/etc. We should do a broader survey of methods to get a scope of the work and the potential payoff for the effort.

### PR DESCRIPTION
**Pull request by Anmol-Srivastava**
Closes #125 

### Added
- functions for injecting drift according to Dirichlet, tweak-one, and minority shift methods
- accompanying tests/documentation

### Details
- excludes other drift injection techniques that may be implemented elsewhere in OS code

### Checklist
- [ ] tests added and passing
- [x] docs added and building
- [x] pandas support added

### CODE REVIEW

File: menelaus/injection/class_manipulation.py [hunk 1]
```diff
@@ -94,3 +96,119 @@ def class_join(data, target_col, class_1, class_2, new_class, from_index, to_ind
...
+    """
+    ret = np.copy(data)
+    classes = np.unique(ret[:, target_col])
+
+    # distribution for each data point, and reordering of each point by class
+    p_distribution = np.array([])
+    sample_grouped = np.array([])
+
+    # locate all classes to shift
+    shifted_classes_idx = np.where(data[:, target_col].isin(classes))[0]
```
Comment (indialindsay) [inline menelaus/injection/class_manipulation.py]: I tried testing out these functions with a dataset on my end to get a better understanding of how they work. I’m encountering a few errors related to the data type. The input for this one is np.array input. I got an error as I think 'isin' only works with pd.dataframes? 

Do you think we should set up this function work with both pd.dataframes and np.arrays? Or just one of them. If you have a busy week and want help adjusting the syntax, let me know.

File: menelaus/injection/class_manipulation.py [hunk 2]
```diff
@@ -94,3 +96,119 @@ def class_join(data, target_col, class_1, class_2, new_class, from_index, to_ind
...
+
+    Args:
+        data (np.array): data to inject with drift
+        target_col (int or str): column index/label of targets column
+        from_index: row index at which to start class swap
+        to_index: row index at which to end (non-inclusive) class swap
+        alpha (dict): used to derive alpha parameter for Dirichlet
+            distribution. Keys are all labels, values are the desired
+            average weight (typically ``int``) per label when resampling. 
+        p_classes (float): altered sampling chance for members of ``classes``
```
Comment (indialindsay) [inline menelaus/injection/class_manipulation.py]: I'm assuming p_classes is in the documentation from copying it over, it's not a paramater right?

**Comment by indialindsay** [comment]
General comment on class join, I know you wrote this prior but took a second look at it. 

Rereading the original DeSouza description of this method, the way they set this method up was to essentially introduce a new class that is not present in the training set. This type of drift injection may cause issues for users hoping to apply our concept drift detection algorithms, since ours are not capable of detecting a new class. Options: 
    1. We could only use this drift injection method for detecting data drift and allow a user to specify a new class label for these joined classes. Then, a data drift detector could be used to detect this new change in the distribution of the class feature. My main concern is this could negate the purpose of this method - a user would ideally use a class joining drift injection technique with a classifier and a concept drift detector. 
    2. We could just mention in the documentation for this algorithm that a user should only assign a label for the newly joined classes that is already present in the training data. Then, it would merge two classes together and it could be used with a learner / concept drift detector. 

Thoughts?
tagging @tms-bananaquit for comments too

File: menelaus/injection/class_manipulation.py (hunk 1 above)
Comment (Anmol-Srivastava) [reply menelaus/injection/class_manipulation.py]: I think they definitely should support both, as the original functions I wrote do. 

It looks like I haven't quite tested even the `numpy` logic yet, but for these new functions I was thinking of just doing a DF => `numpy` => back to DF conversion to handle the `pandas` case.

File: menelaus/injection/class_manipulation.py (hunk 2 above)
Comment (Anmol-Srivastava) [reply menelaus/injection/class_manipulation.py]: Good catch, thanks! Not a parameter for D-shift

**Comment by tms-bananaquit** [comment]
I think the approach in (2) is reasonable, and straightforward to implement with the code as it is now. We could default the `new_class` argument to the same value as `class_1` (or `class_2`, whichever) if it's not specified, and tweak the docstrings to reflect that. 

Right now, `menelaus` only has concept drift detectors that expect binary outcomes. I think the code might actually run for the multinomial case, but (usually) it'd be evaluating overall accuracy (`predicted_class == true_class`), which isn't necessarily desirable in the multinomial case. Inserting a new class label would mean accuracy would immediately tank on the modified classes, though, since the classifier *can't* spit out a label that's not in the training data. In that sense, the detection would probably fire, assuming that the novel class is prevalent enough in new data. 

"Renaming" two classes to the same thing doesn't actually make sense with a binary classifier, unless there's already an additional preprocessing step, when training the classifier, that buckets all the classes into 0/1. I'm not sure we need to worry *too* hard about that one, though that might be a candidate for later experiments.

File: menelaus/injection/class_manipulation.py [hunk 3]
```diff
@@ -94,3 +96,119 @@ def class_join(data, target_col, class_1, class_2, new_class, from_index, to_ind
         class_idx = class_idx[(class_idx < to_index) & (class_idx >= from_index)]
         ret[class_idx, target_col] = new_class
     return ret
+
+# endregion
+
+
+# region - LTF-inspired class manipulation functions
+
+def class_probability_shift(data, target_col, from_index, to_index, classes, p_classes):
```
Comment (indialindsay) [inline menelaus/injection/class_manipulation.py]: Here’s how I’m understanding the purpose of this function: 
Probability_shift receives a float as input and sets the proportion of the specified label to be that float, while other classes are distributed uniform. If multiple labels are specified, it sets the proportion of each of those classes to the specified float, with the rest uniform. 

This method implements the tweak one shift from LTF. If wanting to implement minority-class shift from LTF, you would simply pass a minority class label and desired proportion. 

We might need to add in an error catch statement to make sure p_classes*the number of classes input to ‘classes’ does not exceed 1?

File: menelaus/injection/class_manipulation.py [hunk 4]
```diff
@@ -94,3 +96,119 @@ def class_join(data, target_col, class_1, class_2, new_class, from_index, to_ind
...
+
+    # locate each class in window
+    for cls in classes:
+        cls_idx = np.where(data[:, target_col] == cls)[0]
+        cls_idx = cls_idx[
+            (cls_idx < to_index) &
+            (cls_idx >= from_index)
+        ]
+
+        if cls in classes:
```
Comment (indialindsay) [inline menelaus/injection/class_manipulation.py]: are you meaning to reference the classes variable defined in L127? I think it should be checking if cls in the classes from the input parameter?

File: menelaus/injection/class_manipulation.py (hunk 3 above)
Comment (Anmol-Srivastava) [reply menelaus/injection/class_manipulation.py]: Yeah, my thinking matches the above^

I'll add the check, I was thinking we may need something or another given the different combinations in which this method can be called

File: menelaus/injection/class_manipulation.py (hunk 4 above)
Comment (Anmol-Srivastava) [reply menelaus/injection/class_manipulation.py]: Oops, yes, this should iterate over all classes, and then the `if` statement should refer to the shifted classes, but refers to the modified variable instead.

File: menelaus/injection/class_manipulation.py [hunk 5]
```diff
@@ -94,3 +96,119 @@ def class_join(data, target_col, class_1, class_2, new_class, from_index, to_ind
...
+        sample_grouped = np.concatenate((sample_grouped, data[cls_idx]))
+        p_distribution = np.concatenate(p_distribution, np.ones(cls_idx.shape[0]) * p_individual)
+
+    # shuffled sample over window, with replacement, with weights
+    sample = np.random.choice(sample_grouped, to_index-from_index, True, p_distribution)
+    ret[from_index:to_index] = sample
+    return ret
+
+
+def class_dirichlet_shift(data, target_col, from_index, to_index, alpha):
```
Comment (indialindsay) [inline menelaus/injection/class_manipulation.py:208]: and to the best of my understanding of dirichlet, this one seems implemented correctly too. 
To double check the way I'm understanding it.. 
The general functionality is to change labels so they appear in proportions as determined by a dirichlet distribution. In the input dictionary, users specify parameters for dirchlet for each class.
- Like previous function, we may need to check that the inputted proportions in alpha exactly equals 1

**Review by indialindsay — APPROVED** [review]

File: menelaus/injection/class_manipulation.py (hunk 5 above)
Comment (Anmol-Srivastava) [reply menelaus/injection/class_manipulation.py:208]: Yep, that's the reasoning I used and what I'm thinking the `numpy` function is doing.

File: menelaus/injection/class_manipulation.py (hunk 3 above)
Comment (Anmol-Srivastava) [reply menelaus/injection/class_manipulation.py]: On second thought, should the check just be `if p_classes >= 1 ... raise ...`? @indialindsay I think the correction to the above is that if multiple labels are specified, the probability to sample from any of the shifted classes is `p_classes`. 

So all `classes` share in one specified probability when being sampled, in which case -- does this change your view on the implementation?

Comment (tms-bananaquit) [reply menelaus/injection/class_manipulation.py]: I think the code is treating the parameter a little differently than India's expecting, maybe.

India's description, put another way - if you had five classes, two of which were specified to be sampled with probability `p_classes`, then the remaining 3 categories would need to be resampled, uniform with probability `(1 - 2*p_classes)/3`.

If I'm reading it right, the code's current behavior would be to have new probability of `p_classes/2` for each of the classes you're upsampling (L146), and then `(1 - p_classes)/3` for the rest (L149).

With the latter, you don't need a check. With India's parameterization, you would need something to make sure that `p_classes * len(classes) <= 1` (or something to scale the user's input down to a valid probability vector, which seems a little unclean).

Comment (tms-bananaquit) [reply menelaus/injection/class_manipulation.py]: I'm not sure if I'd find the `p_classes` parameter being reduced by the number of `classes` intuitive as a user.

If this function were more general, with list or dict parameters, could we re-use it? e.g.
```
class_probability_shift(..., foo, ...)
  """
  foo (dict) - dictionary of the form {class, probability} that defines new sampling probabilities. 
                   Undefined classes are resampled uniformly.
  """
  ...

class_dirichlet_shift(...)
  """
  alpha - parameter for Dirichlet distribution ...
  """
  new_vec = dirichlet_distribution(alpha)
  #then set up new_vec as a dictionary for classes...
  class_probability_shift(..., foo=new_vec, ...)
```

Would something like that work? Or have I misunderstood the Dirichlet shift setup?

Comment (indialindsay) [reply menelaus/injection/class_manipulation.py]: It'll be helpful to think through an example to all be on the same page. 

If we have 100 samples and there are 5 possible labels (0-4). The user wants to increase labels 0, 1, and 2 to appear with probability = 0.3, and labels 4, 5 to be uniform (but not disappear). The user passes in p_classes = 0.3 and classes = [1,2,3]. @class_probability_shift should assign:
- 30 samples for label 0 (0.3*total samples)
- 30 for label 1: 0.3*total samples
- 30 for label 2: 0.3*total samples
- 5 for label 4:  (1-0.3*3)*total samples
- 5 for label 5:  (1-0.3*3)*total samples

We would need an error statement to make sure p_classes*len(classes) <= 1. For example, passing in 4 class labels and p_classes = 0.3 would result in a 'error'. 

Let me know if i'm interpreting the function wrong. Or do we want this to function in some other way?

Comment (indialindsay) [reply menelaus/injection/class_manipulation.py]: Then for the dirichlet, that function allows a user to specify different proportions to a dirichlet distribution for different classes. 

For the same example, a user could want {0:0.2, 1:0.3, 2:0.4, 3:0.05, 4:0.05}. So we would need first to check to make sure 0.2 + 0.3 + 0.4 + 0.05 + 0.05 <= 1. We pass alpha = [0.2, 0.3, 0.4, 0.05, 0.05] to the np.dirichlet method and it returns a proportion to sample those classes from the total number of observations in the data, according to the dirichlet distribution. (I believe the np.dirichlet method will not work if the sum(alpha) > 1). 

Following off of your second comment @tms-bananaquit, I don't think we could pass that into @class_proportion_shift since new_vec would contain an array of different proportions, not one proportion as expected by the p_classes parameter.

File: menelaus/injection/class_manipulation.py (hunk 5 above)
Comment (Anmol-Srivastava) [reply menelaus/injection/class_manipulation.py:208]: @indialindsay Since the alpha interpretation I get from `numpy` is an integer weights sort of thing, does this add-to-1 check still apply? I was thinking something like

```
1 + 10 + 2 = 13
proportions are 1, 10, 2 out of 13
```

In which case a check wouldn't make sense -- do you have a different interpretation?

Comment (indialindsay) [reply menelaus/injection/class_manipulation.py:208]: yes you're right, a check doesn't make sense when using the dirichlet function.. think it's all good to not add that in

**Comment by Anmol-Srivastava** [comment]
Oh hm everything is working apparently. I'd say this is ready for a new review @indialindsay, whenever you have time
